patches/server/1055-Adding-PredicateChoice-to-Paper-API.patch

```diff
@@ -0,0 +1,210 @@
+     private final StackedContents contents;
+     public final Map<Ingredient, IntList> extraStackingIds = new IdentityHashMap<>();
++    private Recipe<?> recipe;
+ 
+     public StackedContentsExtraMap(final StackedContents contents, final Recipe<?> recipe) {
+         this.exactChoiceIds.defaultReturnValue(-1);
+         this.contents = contents;
+         this.initialize(recipe);
+     }
++    /**
++     * Instead of using the items an ingredient provides we choose a different approach.
++     * PredicateChoices do not have those "template items". That is why we use the items in the stacked contents as "template items"
++     * Whenever we account a stack we test if there is any ingredient that would return true on test(itemStack).
++     * Else we act as if the item was not in the inventory.
++     * As a result, there is no need to change the vanilla implementation of RecipePicker.
++     */
+ 
+     private void initialize(final Recipe<?> recipe) {
+-        if (recipe.hasExactIngredients()) {
+-            for (final Ingredient ingredient : recipe.getIngredients()) {
+-                if (!ingredient.isEmpty() && ingredient.exact) {
+-                    final net.minecraft.world.item.ItemStack[] items = ingredient.getItems();
+-                    final IntList idList = new IntArrayList(items.length);
+-                    for (final ItemStack item : items) {
+-                        idList.add(this.registerExact(item)); // I think not copying the stack here is safe because cb copies the stack when creating the ingredient
+-                        if (!item.hasTag()) {
+-                            // add regular index if it's a plain itemstack but still registered as exact
+-                            idList.add(StackedContents.getStackingIndex(item));
+-                        }
+-                    }
+-                    idList.sort(IntComparators.NATURAL_COMPARATOR);
+-                    this.extraStackingIds.put(ingredient, idList);
+-                }
+-            }
++        this.recipe = recipe;
++        for (final Ingredient ingredient : recipe.getIngredients()) {
++            if(ingredient.isEmpty())
++                continue;
++            extraStackingIds.put(ingredient, new IntArrayList());
+         }
+     }
+ 
+     private int registerExact(final ItemStack exactChoice) {
++        if(exactChoice.isEmpty())
++            return -1;
+         final int existing = this.exactChoiceIds.getInt(exactChoice);
+         if (existing > -1) {
+             return existing;
+         }
+         final int id = this.idCounter.getAndIncrement();
++        boolean isPotentialIngredient = false;
++        for (final Ingredient ingredient : this.recipe.getIngredients()) {
++            if (ingredient.test(exactChoice)) {
++                isPotentialIngredient = true;
++                final IntList idList = extraStackingIds.computeIfAbsent(ingredient, ingredient1 -> new IntArrayList());
```

Comment:
Should already be in the map right? I think it should throw if its not (or leave this, and don't pre-populate in the initialize method)

Reply:
Yes.